File: playonkodi/playonkodi.py
import requests
import os
import random
import socket
import subprocess
import time
import argparse
import sys

KODI_HOST = '127.0.0.1'
KODI_PORT = 8050


def parse_args():
    parser = argparse.ArgumentParser(
        description='Stream your local/network content directly on Kodi.',
        formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument(
        'media',
        metavar='MEDIA',
        type=str,
        help='path to file containing media')
    return parser


def get_local_ip():
    local = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    local.connect(("8.8.8.8", 80))
    ip = local.getsockname()[0]
    local.close()
    return ip


def parse_filepath(filepath):
    splits = filepath.split('/')
    directory = '/'.join(splits[:-1])
    filename = splits[-1]
    return (filename, directory)


def kodi_post(network_file, KODI_HOST, KODI_PORT):
    data = '{"id":529,"jsonrpc":"2.0","method":"Player.Open","params":{"item":{"file":"' + network_file + '"}}}'
    response = requests.post('http://{0}:{1}/jsonrpc'.format(KODI_HOST, KODI_PORT), data=data)
    return response


def command_line():
    local_ip = get_local_ip()
    local_port = str(random.randint(45000, 50000))

    parser = parse_args()
    raw_args = sys.argv[1:]
    args = parser.parse_args(raw_args)

    filepath = args.media

    if '://' in filepath:
        response = kodi_post(filepath, KODI_HOST, KODI_PORT)
        print(response.text)
    else:
        filename, directory = parse_filepath(filepath)

        command = ['python', 'server.py', directory, local_port]
        server = subprocess.Popen(
                      command,
                      shell=False,
                      stdout=subprocess.PIPE,
                      stderr=subprocess.PIPE)

        network_file = 'http://{0}:{1}/content/{2}'.format(local_ip, local_port, filename)
        print(network_file)

        time.sleep(2)

        response = kodi_post(network_file, KODI_HOST, KODI_PORT)
        print(response.text)

        print('\nHit CTRL+C to kill the stream webserver')

        try:
            while server.poll() is None:
                time.sleep(1)
        except KeyboardInterrupt:
            server.terminate()
            raise


if __name__ == '__main__':

    command_line()

File: bin/playonkodi.js
#!/usr/bin/env node

var ArgumentParser = require('argparse').ArgumentParser;
var http = require('http');
var request = require('request');

var finalhandler = require('finalhandler');
var serveStatic = require('serve-static');
var path = require('path');
var os = require('os');
var exec = require('child_process').exec;


function parse_args() {
    var parser = new ArgumentParser({
        version: '0.2.0',
        addHelp: true,
        description: 'Stream your local/network content directly on Kodi.',
    });
    parser.addArgument(
        [ 'media' ], {
        metavar: 'MEDIA',
        type: 'string',
        help: 'Path to media file'
        }
    );
    parser.addArgument(
        [ '-s', '--server' ], {
        type: 'string',
        required: true,
        help: "Kodi's local ip address"
        }
    );
    parser.addArgument(
        [ '-p', '--port' ], {
        type: 'int',
        required: true,
        help: "Kodi's web interface port"
        }
    );
    parser.addArgument(
        [ '-i', '--interface-ip' ], {
        type: 'string',
        help: "[Optional] Interface IP to send to Kodi server"
        }
    );
    parser.addArgument(
        [ '--no-resolve' ], {
        action: 'storeTrue',
        help: "[Optional] Do not resolve the external media URL with youtube-dl"
        }
    );
    return parser;
}


function get_localip() {
    var ifaces = os.networkInterfaces();
    var localip = '';

    Object.keys(ifaces).forEach(function (ifname) {
        ifaces[ifname].forEach(function (iface) {
            if ('IPv4' !== iface.family || iface.internal !== false) {
                // skip over internal (i.e. 127.0.0.1) and non-ipv4 addresses
                return;
            }
            localip = iface.address;
        });
    });

    return localip;
}


function serve_directory(directory, port) {
    var serve = serveStatic(directory);

    var server = http.createServer(function(req, res) {
        var done = finalhandler(req, res);
        serve(req, res, done);
    });

    server.listen(port);
}


function kodi_post(network_file, server, port) {
    console.log('Commanding jsonrpc on ' + server_ip + ':' + server_port + ' to listen for media content on the resolved URL');
    var dataString = '{"id":519,"jsonrpc":"2.0","method":"Player.Open","params":{"item":{"file":"' + network_file + '"}}}';

    var options = {
        url: 'http://' + server + ':' + port + '/jsonrpc',
        method: 'POST',
        body: dataString
    };

    console.log('The media content should play now');
    return request(options);
}

function execute(command, callback) {
    exec(command, function(error, stdout, stderr) {
        callback(stdout);
    });
};

var parser = parse_args();
var args = parser.parseArgs();

const filepath = args.media
const server_ip = args.server
const server_port = args.port


if (filepath.indexOf('://') == -1) {
    var local_ip = args.interface_ip;
    if (local_ip == null) {
        local_ip = get_localip();
    }
    const local_port = '15000';
    console.log(local_ip);

    var directory = path.dirname(filepath);
    var network_file = 'http://' + local_ip + ':' + local_port + '/' + encodeURIComponent(path.basename(filepath));

    console.log('Hosting media content on:')
    console.log(network_file + '\n');

    serve_directory(directory, local_port);
    kodi_post(network_file, server_ip, server_port);

    console.log('\nHit Ctrl+C to kill the local stream server');

} else {
    var to_resolve = !args.no_resolve;
    if (to_resolve) {
        console.log('Resolving URL using youtube-dl');
        execute('youtube-dl -gf best ' + filepath, function(resolved) {
            resolved = resolved.replace('\n', '');
            kodi_post(resolved, server_ip, server_port);
        });
    } else {
        kodi_post(filepath, server_ip, server_port);
    }
}
